### PR TITLE
Support for Microchip Same51

### PR DESCRIPTION
**Pull request by danielinux**
- [x] hal support
- [x] Dual-bank swapping
- [x] update Target.md
- [x] Add build test

### CODE REVIEW

File: .gdbinit [hunk 1]
```diff
@@ -1,6 +1,7 @@
 tar rem:3333
 file wolfboot.elf
-add-symbol-file test-app/image.elf
+#add-symbol-file test-app/image.elf
+add-symbol-file /home/dan/src/wolfboot/IDE/MPLabX/wolfBoot_usbUpdateApp.X/dist/default/production/wolfBoot_usbUpdateApp.X.production.elf
```
Comment (dgarske) [inline .gdbinit:4]: Consider reverting this.

File: IDE/MPLabX/wolfBoot-SAME51.X/Makefile [hunk 2]
```diff
@@ -0,0 +1,113 @@
+#
```
Comment (dgarske) [inline IDE/MPLabX/wolfBoot-SAME51.X/Makefile:1]: Please consider using `IDE/MPLABX` to match our other libraries and Microchip naming.

File: hal/same51.c [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+    for (r = 0; r < len; r += WOLFBOOT_SECTOR_SIZE) {
+        hal_flash_write(BANKB_BASE + r, (void *)(BANKA_BASE + r), WOLFBOOT_SECTOR_SIZE);
+    }
+    hal_flash_lock();
+}
+
+void hal_init(void)
+{
+
+    uint32_t reg;
```
Comment (dgarske) [inline hal/same51.c:280]:
```hal/same51.c: In function 'hal_init':
hal/same51.c:280:14: error: unused variable 'reg' [-Werror=unused-variable]
  280 |     uint32_t reg;
      |              ^~~
```

File: hal/same51.c [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+        ;
+
+}
+
+#define BANKA_BASE 0x00000000
+#define BANKB_BASE 0x00080000
+
+#include <string.h>
+
+static void RAMFUNCTION fork_bootloader(void)
```
Comment (dgarske) [inline hal/same51.c:263]: Consider wrapping this in `DUALBANK_SWAP`.

```
hal/same51.c:263:25: error: 'fork_bootloader' defined but not used [-Werror=unused-function]
  263 | static void RAMFUNCTION fork_bootloader(void)
      |                         ^~~~~~~~~~~~~~~```

File: IDE/MPLabX/wolfBoot-SAME51.X/Makefile [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+#  There exist several targets which are by default empty and which can be
```
Comment (dgarske) [inline IDE/MPLabX/wolfBoot-SAME51.X/Makefile:2]: Please add a README.md for how to use these MPLABX projects.

File: hal/same51.c (hunk 4 above)
Comment (dgarske) [reply hal/same51.c:263]: I added `CFLAGS+=-Wunused -Wunused-parameter` to get these.

**Review by dgarske — CHANGES_REQUESTED** [review]

**Comment by danielinux** [comment]
Thanks for the review, @dgarske . I should have addressed the comments + re-created the two MPLAB projects + provided extended documentation. This is now ready to consider for merging.

**Pull request merged**